### PR TITLE
Create a new TyKind::DynStar variant

### PR DESCRIPTION
**Pull request by eholk**
Previously `dyn*` was implemented by adding a `DynKind` field on the `ty::Dynamic` variant. This works well for cases where we can treat the two types mostly the same, which is most of the front parts of the compiler (type checker, etc.). It doesn't work so well for the backends like Miri, since the representation of a `dyn` and `dyn*` are quite different. This PR splits the `dyn*` type into its own variant to make it more explicit how it is handled.

r? compiler-errors

### CODE REVIEW

**Comment by eholk** [comment]
This is almost certainly going to break in CI... I fixed what I could test locally, but unfortunately I haven't found a great way to test things like cranelift on my own machine.

**Comment by RalfJung** [comment]
This will also cause tons of conflicts with https://github.com/rust-lang/rust/pull/107728 :/

**Comment by compiler-errors** [comment]
Needs fixing, and probably should wait until Ralf's PR is landed.

@rustbot author

**Comment by JohnCSimon** [comment]
@eholk 
ping from triage - can you post your status on this PR?




> This will also cause tons of conflicts with #107728 :/

this has merged


FYI: when a PR is ready for review, send a message containing 
`@rustbot ready` to switch to `S-waiting-on-review` so the PR is in the reviewer's backlog.

File: src/tools/clippy/clippy_lints/src/ptr.rs [hunk 1]
```diff
@@ -1,19 +1,23 @@
 //! Checks for usage of  `&Vec[_]` and `&String`.
 
-use clippy_utils::diagnostics::{span_lint, span_lint_and_sugg, span_lint_and_then, span_lint_hir_and_then};
+use clippy_utils::diagnostics::{
```
Comment (flip1995) [inline src/tools/clippy/clippy_lints/src/ptr.rs:3]: This file looks as if it was just re-formatted. Please don't commit formatting changes to Clippy in the Rust repo.

**Comment by compiler-errors** [comment]
Gonna close this since it's pretty stale, and probably just needs to be done from scratch.

**Comment by RalfJung** [comment]
> it's pretty stale, and probably just needs to be done from scratch.

Where is that tracked?

**Comment by compiler-errors** [comment]
@RalfJung, what do you mean?

**Comment by RalfJung** [comment]
Do we have an issue tracking this cleanup and the reasons for it? Chances are it will just get forgotten otherwise.

I guess the tracking issue for dynstar would be the right place?

**Comment by compiler-errors** [comment]
Ah, no, we currently do not have one. I'll leave a comment on the dyn* tracking issue.

**Pull request closed without merging**